cmd/testscript/main.go

```diff
@@ -0,0 +1,219 @@
+	verbose bool
+}
+
+func (r runner) Skip(is ...interface{}) {
+	panic(skipRun)
+}
+
+func (r runner) Fatal(is ...interface{}) {
+	r.Log(is...)
+	r.FailNow()
+}
+
+func (r runner) Parallel() {
+	// No-op for now; we are currently only running a single script in a
+	// testscript instance.
+}
+
+func (r runner) Log(is ...interface{}) {
+	fmt.Print(is...)
+}
+
+func (r runner) FailNow() {
+	panic(failedRun)
+}
+
+func (r runner) Run(n string, f func(t testscript.T)) {
+	// For now we we don't top/tail the run of a subtest. We are currently only
+	// running a single script in a testscript instance, which means that we
+	// will only have a single subtest.
+	f(r)
+}
+
+func (r runner) Verbose() bool {
+	return r.verbose
+}
+
+func run(runDir, fileName string, verbose bool) error {
+	var ar *txtar.Archive
+	var err error
+
+	mods := filepath.Join(runDir, goModProxyDir)
+
+	if err := os.MkdirAll(mods, 0777); err != nil {
+		return fmt.Errorf("failed to create goModProxy dir: %v", err)
+	}
+
+	if fileName == "-" {
+		fileName = "<stdin>"
+		byts, err := ioutil.ReadAll(os.Stdin)
+		if err != nil {
+			return fmt.Errorf("failed to read from stdin: %v", err)
+		}
+		ar = txtar.Parse(byts)
+	} else {
+		ar, err = txtar.ParseFile(fileName)
+	}
+
+	if err != nil {
+		return fmt.Errorf("failed to txtar parse %v: %v", fileName, err)
+	}
+
+	var script, gomodProxy txtar.Archive
+	script.Comment = ar.Comment
+
+	for _, f := range ar.Files {
+		fp := filepath.Clean(filepath.FromSlash(f.Name))
+		parts := strings.Split(fp, string(os.PathSeparator))
+
+		if len(parts) > 1 && parts[0] == goModProxyDir {
+			gomodProxy.Files = append(gomodProxy.Files, f)
+		} else {
+			script.Files = append(script.Files, f)
+		}
+	}
+
+	if txtar.Write(&gomodProxy, runDir); err != nil {
+		return fmt.Errorf("failed to write .gomodproxy files: %v", err)
+	}
+
+	if err := ioutil.WriteFile(filepath.Join(runDir, "script.txt"), txtar.Format(&script), 0666); err != nil {
+		return fmt.Errorf("failed to write script for %v: %v", fileName, err)
+	}
+
+	p := testscript.Params{
+		Dir: runDir,
+	}
+
+	if len(gomodProxy.Files) > 0 {
+		srv, err := goproxytest.NewServer(mods, "")
+		if err != nil {
+			return fmt.Errorf("cannot start proxy for %v: %v", fileName, err)
+		}
+		defer srv.Close()
+
+		currSetup := p.Setup
+
+		p.Setup = func(env *testscript.Env) error {
+			env.Vars = append(env.Vars, "GOPROXY="+srv.URL)
+			if currSetup != nil {
+				return currSetup(env)
+			}
+			return nil
+		}
+	}
+
+	if err := gotooltest.Setup(&p); err != nil {
+		return fmt.Errorf("failed to setup go tool for %v run: %v", fileName, err)
+	}
+
+	r := runner{
+		verbose: verbose,
+	}
+
+	func() {
+		defer func() {
+			switch recover() {
+			case nil, skipRun:
+			case failedRun:
+				err = failedRun
+			default:
+				panic(fmt.Errorf("unexpected panic: %v [%T]", err, err))
+			}
+		}()
+		testscript.RunT(r, p)
+	}()
+
+	if err != nil {
+		return fmt.Errorf("failed to run test for %v in %v\n", fileName, runDir)
```

Comment:
Nit: we managed to *run* the test, but it failed. How about:

    return fmt.Errorf("error running %v in %v", fileName, runDir)

?

Reply:
Good point, thanks. Fixed